rfcs/12-primitive-shapes.md

```diff
@@ -0,0 +1,554 @@
+}
+```
+### 2D Geometry Types
+
+```rust
+struct Direction2d(Vec2)
+impl Meshable for Direction2d {}
+
+struct Ray2d {
+  point: Vec2, 
+  direction: Direction2d
+}
+impl Meshable for Ray2d {}
+
+struct Line2d {
+  point: Vec2, 
+  direction: Direction2d,
+}
+impl Meshable for Line2d {}
+
+/// A finite line between two vertices
+struct LineSegment2d { 
+  start: Vec2, 
+  end: Vec2,
+}
+impl Meshable for LineSegment2d {}
+
+/// A line drawn along a path of vertices 
+struct PolyLine2d<const N: usize>{
+  points: [Vec2; N],
+}
+impl Meshable for PolyLine2d {}
+
+struct Triangle2d([Vec2; 3]);
+impl Meshable for Triangle2d {}
+
+struct Quad2d([Vec2; 4]);
+impl Meshable for Quad2d {}
+
+/// A closed shape in a plane defined by a finite number of line-segments between vertices
+struct Polygon2d <const N: usize>{
+  points: [Vec2; N],
+}
+impl Meshable for Polygon2d {}
+
+/// A polygon where all vertices lies on the circumscribed circle, equally far apart
+/// Example: Square, Triangle, Hexagon
+struct RegularPolygon2d {
+  /// The circumcircle that all points of the regular polygon lie on.
+  circumcircle: Circle2d,
+  /// Number of faces.
+  faces: u8,
+  /// Clockwise rotation of the polygon about the origin. At zero rotation, a point will always be located at the 12 o'clock position.
+  orientation: f32,
+}
+impl Meshable for RegularPolygon2d {}
+
+
+// Circle types
+
+/// A circle defined by its radius
+struct Circle {
+  radius: f32,
+}
+impl Meshable for Circle {}
+
+/* REFERENCE ONLY
+struct BoundingCircle2d {
+  circle: Circle,
+  translation: Vec2,
+}
+impl Meshable for BoundingCircle2d {}
+impl Bounding2d for BoundingCircle2d {}
+
+struct CircleCollider {
+  sphere: Circle,
+  translation: Vec2,
+}
+impl Meshable for CircleCollider {}
+impl Collider2d for CircleCollider {}
+*/
+
+// Box Types
+
+/// A rectangle defined by its width and height
+struct Rectangle {
+  half_extents: Vec2,
+}
+impl Meshable for Rectangle
+
+/* REFERENCE ONLY
+struct BoundingBox2d {
+  box: Rectangle,
+  translation: Vec2,
+}
+impl Meshable for BoundingBox2d {}
+impl Bounding2d for BoundingBox2d {}
+type Aabb2d = BoundingBox2d;
+
+struct RectangleCollider {
+  box: Rectangle,
+  translation: Vec2,
+  rotation: Mat2,
+}
+impl Meshable for RectangleCollider {}
+impl Collider2d for RectangleCollider {}
+type Obb2d = RectangleCollider;
+*/
+
+// Capsule Types
+
+struct Capsule2d {
+  height: f32, // Height of the rectangular section
+  radius: f32, // End cap radius
+}
+impl Meshable for Capsule2d {}
+
+/* REFERENCE ONLY
+struct CapsuleCollider2d {
+  capsule: Capsule2d,
+  translation: Vec2,
+  rotation: Mat2,
+}
+impl Meshable for CapsuleCollider2d {}
+impl Collider2d for CapsuleCollider2d {}
+*/
+
+```
+
+### 3D Geometry Types
+
+```rust
+/// Vector direction in 3D space that is guaranteed to be normalized through its getter/setter.
+struct Direction3d(Vec3)
+impl Meshable for Direction3d {}
+
+struct Plane3d {
+  point: Vec3,
```

Comment:
Should those be `Vec3A` instead of `Vec3`? If the focus is performance, surely we'd benefit from SIMD usage no?

Reply:
The implementation of the shapes are out of scope of this RFC, but the way the suggested api works we could easily have both variants. Example, by having a `Plane3d` and a `Plane3dA`